test/boost/sstable_compaction_test.cc

```diff
@@ -5005,38 +5006,40 @@ SEASTAR_TEST_CASE(cleanup_incremental_compaction_test) {
                 std::move(mut2)
             });
             sstables::test(sst).set_run_identifier(run_identifier); // in order to produce multi-fragment run.
-            sst->set_sstable_level(1);
+            // Set level to 0 to trigger offstrategy compaction
+            sst->set_sstable_level(offstrategy ? 0 : 1);
 
             // every sstable will be eligible for cleanup, by having both an owned and unowned token.
             owned_token_ranges.push_back(dht::token_range::make_singular(sst->get_last_decorated_key().token()));
 
             gens.insert(sst->generation());
             ssts.push_back(std::move(sst));
+
+            if (offstrategy) {
+                // Force a new run_id to trigger offstrategy compaction
+                run_identifier = run_id::create_random_id();
+            }
         }
 
-        size_t last_input_sstable_count = sstables_nr;
         {
             auto t = env.make_table_for_tests(s);
             auto& cm = t->get_compaction_manager();
             auto stop = deferred_stop(t);
             t->disable_auto_compaction().get();
             const dht::token_range_vector empty_owned_ranges;
             for (auto&& sst : ssts) {
-                testlog.info("run id {}", sst->run_identifier());
-                column_family_test(t).add_sstable(sst).get();
+                testlog.info("run id {}, offstrategy={}", sst->run_identifier(), offstrategy);
+                column_family_test(t).add_sstable(sst, offstrategy).get();
                 column_family_test::update_sstables_known_generation(*t, sst->generation());
                 observers.push_back(sst->add_on_closed_handler([&] (sstable& sst) mutable {
                     auto sstables = t->get_sstables();
-                    auto input_sstable_count = std::count_if(sstables->begin(), sstables->end(), [&] (const shared_sstable& sst) {
-                        return gens.count(sst->generation());
-                    });
-
-                    testlog.info("Closing sstable of generation {}, table set size: {}", sst.generation(), input_sstable_count);
+                    testlog.info("Closing sstable of generation {}, table set size: {}", sst.generation(), sstables->size());
                     sstables_closed++;
-                    if (std::cmp_less(input_sstable_count, last_input_sstable_count)) {
-                        sstables_closed_during_cleanup++;
-                        last_input_sstable_count = input_sstable_count;
-                    }
+                }));
+                observers.push_back(sst->add_on_delete_handler([&] (sstable& sst) mutable {
+                    auto missing = !file_exists(sst.get_filename()).get();
+                    testlog.info("Deleting sstable of generation {}: missing={}", sst.generation(), missing);
+                    sstables_missing_on_delete += missing;
                 }));
```

Comment:
It seems that the test gained the ability to stress the bug in #14304, but lost the ability to check incremental compaction applied during cleanup.

It's not enough to check that we closed all input sstables, but that we closed them as they get exhausted, one by one.

I'd suggest we actually split the tests, each focusing on stressing a particular bug

Reply:
okay, I'll split the cases then

Reply:
one problem though, is that the change defers closure of the sstable till on_compaction_completion so they all are closed in one shot instead of one by one.

Reply:
For keeping the incremental retirement of sstables, we should keep the registration method and just add the cleanup handling in on_compaction_completion on top of it.
I'll do that in v2